File: Dynamic_Programming/decode-ways.cpp
class Solution {
public:
    vector<int>dp;
    
    int decoder(int n,string s)
    {
        if(n>=s.size()) return 1;
        else if(s[n] == '0') return 0;
        else if(n == s.size()-1) return 1;
        
        else if(dp[n] != -1) return dp[n];
        
        else if(s[n] == '1' || (s[n] == '2' && (s[n+1] >= 48 && s[n+1] <= 54)))
        {
            return dp[n] = decoder(n+1,s) + decoder(n+2,s);
        }
        else return dp[n] = decoder(n+1,s);
        
        
    }
    
    
    int numDecodings(string s) {
       dp.resize(s.size());
       dp.assign(s.size(),-1);
        
       return decoder(0,s);
        
        
    }
};

File: Dynamic_Programming/decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [0]*(n+1)
        
        if s[0] == '0':
            return 0
        dp[0] = dp[1] = 1
        
        for i in range(2,len(s)+1):
            if int(s[i-1]) > 0:
                dp[i] = dp[i-1]
            if s[i-2] == '1' or (s[i-2] == '2' and int(s[i-1]) < 7):
                dp[i] += dp[i-2]
        return dp[n]
                
